cpp/src/binaryop/compiled/util.cpp

```diff
@@ -45,7 +45,7 @@ struct common_type_functor {
         // Eg. d=t-t
         return data_type{type_to_id<TypeCommon>()};
       }
-      return {};
+      return std::nullopt;
```

Comment:
A default-constructed `std::optional` should be equivalent to `std::nullopt` from my reading of [cppreference](https://en.cppreference.com/w/cpp/utility/optional/optional). This pattern is used several times in libcudf.

Reply:
I found this: https://stackoverflow.com/questions/57964217/stdoptional-construct-empty-with-or-stdnullopt

Reply:
Here is my reasoning:
 * Implicitly defined (by the compiler) default constructor of a class does not initialize members of built-in types. (https://stackoverflow.com/questions/2417065/does-the-default-constructor-initialize-built-in-types)
 * The class `data_type` has a default constructor, and calling `{}` will construct a default `data_type` object instead of an empty optional object.
 * "a default `data_type` object" does not initialize its members. That's why we got the compile error.

Reply:
> The old code `return {}` will return a valid optional object with default-constructed `data_type`, not a `nullopt`.

I don't think so, mate. [(1) in the spec](https://en.cppreference.com/w/cpp/utility/optional/optional) indicates otherwise. Here's a trivial [godbolt example](https://godbolt.org/z/WfdGPfz85).

This sounds very much like a compiler bug that we're working around. That said, this change might still be worth having.

Reply:
Yeah, the only thing I can think of is a compiler bug that causes `return {}` to return a valid optional object with default-constructed `data_type`.

Reply:
If this is needed as a workaround for GCC 10, I'm not opposed to it -- but we should either make the codebase consistent about using `std::nullopt` rather than the default constructor, OR we should add a comment to indicate this is working around a bug in GCC 10 with the default constructor. I think the workaround-comment is more appropriate in this case.

Reply:
This is also worth to learn: We should prefer EXPLICIT and avoid VAGUE implementation.

Reply:
```suggestion
      // A GCC 10 bug causes a compilation error when using the default
      // constructor of std::optional to indicate a null value. Returning
      // std::nullopt is equivalent to `return {};`
      return std::nullopt;
```

Reply:
As an aside, I'm trying to reason out why this function opts for `std::optional<data_type>` when `type_id` already provides for an empty value (i.e. `type_id::EMPTY`). 
So now, we'd need to protect against `std::nullopt` and `data_type{type_id::EMPTY}`. :/

Reply:
As I reasoned above, although `return {}` may return the correct optional object (containing no value), `{}` makes the compiler to construct a default-value `data_type` anyway (bug here), which has its member not initialized.

Reply:
@ttnghia As an aside, the example of `*ptr++` vs. `(*ptr)++` is explained by [operator precedence](https://en.cppreference.com/w/cpp/language/operator_precedence), where postfix-increment is applied before indirection (dereference).

Reply:
Personally, `return {}` seems to be very confusing to me. Since `std::optional<T>` can be constructed by copy or move an existing object, I am not sure if `{}` will construct an empty optional object, or construct a `T` object then copy/move construct an optional object from that `T` object.
If we have `return std::optional<T>{}` then everything is clear.

Reply:
> @ttnghia As an aside, the example of `*ptr++` vs. `(*ptr)++` is explained by [operator precedence](https://en.cppreference.com/w/cpp/language/operator_precedence), where postfix-increment is applied before indirection (dereference).

Yeah, but I can't remember such operator order. Or even if I remember, I am not sure 100% about such order from my memory until I look up such documentation. So, it is not explicit/trivial to understand :smile:

Reply:
> `{}` makes the compiler to construct a default-value `data_type` object anyway (bug here?)

That disagrees with @mythrocks' example here, which does not call the constructor of the underlying type: https://godbolt.org/z/WfdGPfz85

Reply:
The example above can't reproduce the issue here. Even with the cudf code, the issue only appear if we have the full function:
```
template <typename TypeLhs, typename TypeRhs>
  struct nested_common_type_functor {
    template <typename TypeOut>
    std::optional<data_type> operator()()
    {
      // If common_type exists
      if constexpr (cudf::has_common_type_v<TypeOut, TypeLhs, TypeRhs>) {
        using TypeCommon = typename std::common_type<TypeOut, TypeLhs, TypeRhs>::type;
        return data_type{type_to_id<TypeCommon>()};
      } else if constexpr (cudf::has_common_type_v<TypeLhs, TypeRhs>) {
        using TypeCommon = typename std::common_type<TypeLhs, TypeRhs>::type;
        // Eg. d=t-t
        return data_type{type_to_id<TypeCommon>()};
      }
      return {};
    }
  };
```

If I remove some code to have this, then I'll not have such compilation error:
```
template <typename TypeLhs, typename TypeRhs>
  struct nested_common_type_functor {
    template <typename TypeOut>
    std::optional<data_type> operator()()
    {     
       return {};
    }
  };
```

Reply:
This thread https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80635 is a good starting point for this class of warning in std::optional.

The issue in short is how the `maybe-uninitialized` warning detection behaves, and less about if the code is C++ standard complaint.

The other optional is to disable this warning entirely as it is flaky

Reply:
@ttnghia Can you try moving this function out of the anonymous namespace and see if it compiles on GCC 10 with `return {}`? I think this might be the bug: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=86465

Reply:
Hah, so it's compiler bug not just with gcc-10, but gcc <= 10?

Reply:
> The other **optional** is ...

:] I see what you did there.

Reply:
> Hah, so it's compiler bug not just with gcc-10, but gcc <= 10?

We don't see this bug with GCC 9 or 11. There may be some confluence of issues here that cause it to only appear on GCC 10.

Reply:
It's warning treated as error. From the links above, I can see that such warning is issued with gcc-7 and 8 too.

Reply:
Guys, moving the function out of anonymous namespace fix it!

Reply:
Guys, we have a very hot discussion. Thanks all for your attention :+1: :smiley:

Reply:
> @ttnghia Can you try moving this function out of the anonymous namespace and see if it compiles on GCC 10 with `return {}`?

Moving the function out of anonymous namespace does not help.

Reply:
Nothing to get us excited like a possible compiler bug!